llvm/lib/Support/PropertySetIO.cpp

```diff
@@ -0,0 +1,217 @@
+      Expected<std::unique_ptr<byte[]>> DecArr =
+          Base64::decode(Val.data(), Val.size());
+      if (!DecArr)
+        return DecArr.takeError();
+      Prop.set(DecArr.get().release());
+      break;
+    }
+    default:
+      return createStringError(std::error_code{},
+                               "unsupported property type: ", Ttag.get());
+    }
+    (*CurPropSet)[Parts.first] = std::move(Prop);
+  }
+  if (!CurPropSet)
+    return makeError("invalid property set registry");
+
+  return Expected<std::unique_ptr<PropertySetRegistry>>(std::move(Res));
+}
+
+namespace llvm {
+// output a property to a stream
+raw_ostream &operator<<(raw_ostream &Out, const PropertyValue &Prop) {
+  Out << static_cast<int>(Prop.getType()) << "|";
+  switch (Prop.getType()) {
+  case PropertyValue::Type::UINT32:
+    Out << Prop.asUint32();
+    break;
+  case PropertyValue::Type::BYTE_ARRAY: {
+    util::PropertyValue::SizeTy Size = Prop.getRawByteArraySize();
+    Base64::encode(Prop.asRawByteArray(), Out, (size_t)Size);
+    break;
+  }
+  default:
+    llvm_unreachable(
+        ("unsupported property type: " + utostr(Prop.getType())).c_str());
+  }
+  return Out;
+}
+} // namespace llvm
+
+void PropertySetRegistry::write(raw_ostream &Out) const {
+  for (const auto &PropSet : PropSetMap) {
+    Out << "[" << PropSet.first << "]\n";
+
+    for (const auto &Props : PropSet.second) {
+      Out << Props.first << "=" << Props.second << "\n";
+    }
+  }
+}
+
+namespace llvm {
+namespace util {
+
+template <> uint32_t &PropertyValue::getValueRef<uint32_t>() {
+  return std::get<uint32_t>(Val);
+}
+
+template <> byte *&PropertyValue::getValueRef<byte *>() {
+  auto &ByteArrayVal = std::get<byte *>(Val);
+  return ByteArrayVal;
+}
+
+template <> PropertyValue::Type PropertyValue::getTypeTag<uint32_t>() {
+  return UINT32;
+}
+
+template <> PropertyValue::Type PropertyValue::getTypeTag<byte *>() {
+  return BYTE_ARRAY;
+}
+
+PropertyValue::PropertyValue(const byte *Data, SizeTy DataBitSize) {
+  constexpr int ByteSizeInBits = 8;
+  Ty = BYTE_ARRAY;
+  SizeTy DataSize = (DataBitSize + (ByteSizeInBits - 1)) / ByteSizeInBits;
+  constexpr size_t SizeFieldSize = sizeof(SizeTy);
+
+  // Allocate space for size and data.
+  Val = new byte[SizeFieldSize + DataSize];
+
+  // Write the size into first bytes.
+  for (size_t I = 0; I < SizeFieldSize; ++I) {
+    auto ByteArrayVal = std::get<byte *>(Val);
+    ByteArrayVal[I] = (byte)DataBitSize;
+    DataBitSize >>= ByteSizeInBits;
+  }
+  // Append data.
+  auto ByteArrayVal = std::get<byte *>(Val);
+  std::memcpy(ByteArrayVal + SizeFieldSize, Data, DataSize);
+}
+
+PropertyValue::PropertyValue(const PropertyValue &P) { *this = P; }
+
+PropertyValue::PropertyValue(PropertyValue &&P) { *this = std::move(P); }
+
+PropertyValue &PropertyValue::operator=(PropertyValue &&P) {
+  copy(P);
+
+  if (P.getType() == BYTE_ARRAY) {
+    auto &ByteArrayVal = std::get<byte *>(P.Val);
```

Comment:
```suggestion
    auto *ByteArrayVal = std::get<byte *>(P.Val);
```

Reply:
We need the reference here as we want to set it to nullptr.

Reply:
Modified code to access P.Val directly. 

Thanks